bac-1778: add mtls example in java

renatolean opened the pull request:
Add a new folder containing a way to make call to the lean API using Java.

There are two options available to make the call:

- apache client that loads the certificate in memory
- rest template (from spring boot) that expects a pkcs12 file to be available to connect securely

mostafaleantech committed:
Java branch

renatolean committed:
bac-1778: add mtls example in java

Add a new folder for java content, including a README.md

renatolean committed:
bac-1778: small formatting fix for gitignore

mostafaleantech committed:
adding the pom file

mostafaleantech committed:
Modifying the Main class name

renatolean committed:
bac-1778: update readme

renatolean committed:
Merge branch 'java_branch' into renatosilva/bac-1778-client-side-mtls-example

renatolean committed:
bac-1778: delete unused class

renatolean committed:
bac-1778: test call to sandbox

renatolean committed:
bac-1778: fix typo in application.properties

renatolean committed:
bac-1778: update properties

renatolean committed:
fix typo in base url for prod

mostafaleantech committed:
Adding the documentation

mostafaleantech committed:
removing a *

renatolean committed:
programmatically create keystore to call lean api

Using a helper library "io.github.hakky54", load the certificates in memory to make a TLS call to Lean API

renatolean committed:
rename method

mostafaleantech committed:
removing un-used prop

renatolean committed:
small refactory

renatolean committed:
Merge branch 'java_branch' into renatosilva/bac-1778-client-side-mtls-example

renatolean committed:
update properties file

renatolean committed:
improve wording

renatolean committed:
make compatible with Java 8

renatolean committed:
update readme

renatolean committed:
update gitignore file

renatolean committed:
update readme

renatolean committed:
update readme

illialean commented on mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java:
```diff
@@ -0,0 +1,77 @@
+        try {
+            getBanks();
+        } catch (IOException e) {
+            e.printStackTrace();
```
log.error :P

renatolean replied:
yep, let me fix that!

illialean commented on mtls-examples/java/src/main/java/me/lean/tech/dev/apacheclient/MtlsUsingApacheClient.java:
```diff
@@ -0,0 +1,96 @@
+        PrivateKey privateKey = PemUtils.loadPrivateKey(privateKeyPem);
+
+        KeyStore keyStore = KeyStore.getInstance("pkcs12");
+        keyStore.load(null, null);
```
this looks suspicious, is it expected?

renatolean replied:
yeah, it is expected, see https://www.baeldung.com/java-keystore. Will add a comment to help the reader

illialean commented on mtls-examples/java/pom.xml:
```diff
@@ -0,0 +1,63 @@
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.example</groupId>
```
me.leantech.exaple?

illialean commented on mtls-examples/java/src/main/java/me/lean/tech/dev/apacheclient/MtlsUsingApacheClient.java:
```diff
@@ -0,0 +1,96 @@
+        HttpClientBuilder httpClientBuilder = HttpClientBuilder.create().setSSLContext(sslContext);
+
+        try (CloseableHttpClient httpClient = httpClientBuilder.build()) {
+            String baseUrl = props.getProperty("base.url");
```
do you mention base.url anywhere?

renatolean replied:
yes, on line below, 69

illialean commented on mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java:
```diff
@@ -0,0 +1,77 @@
+    RestTemplate restTemplate(RestTemplateBuilder preconfiguredBuilder) {
+        return preconfiguredBuilder
+                .additionalInterceptors(new LoggingInterceptor())
+                .rootUri("https://api.leantech.me")
```
base.url?

illialean commented on mtls-examples/java/src/main/java/me/lean/tech/dev/springboot/MtlsUsingRestTemplate.java:
```diff
@@ -0,0 +1,77 @@
+    // generic get method to call the bank with the correct headers and return the correct required object class
+    private <T> void makeGetCall(String urlPath, Class<T> responseClass) throws IOException {
+        Properties props = new Properties();
+        props.load(MtlsUsingApacheClient.class.getClassLoader().getResourceAsStream("application.properties"));
```
you are using spring already, maybe inject with value?

renatolean replied:
yeah, good shout!

renatolean committed:
Address code review comments

- Make usage of autowiring in spring call.
- Replace system.out with logs (via lombok)
- Small refactorings

renatolean committed:
add additional troubleshooting cases

renatolean committed:
small formatting fix

renatolean committed:
make application.properties template values consistent

renatolean committed:
add "build" section to the README.md
